packages/extension-commerce-bm-seo/src/setup-app.ts

```diff
@@ -0,0 +1,166 @@
+            parameters: {urlSegment: locals.originalUrl}
+        })
+        console.log('--- url mapping', urlMapping)
+
+        if (!urlMapping.resourceType) {
+            return Promise.resolve([])
+        }
+
+        // Transform URL mapping to serialized route
+        const requestURL = new URL(locals.originalUrl, getAppOrigin(locals))
+        const componentName =
+            config.resourceTypeToComponentMap[urlMapping.resourceType as keyof typeof config.resourceTypeToComponentMap]
+        
+        // Create a placeholder component so it can be deserialized later in beforeRouteMatch
+        const component = Object.assign(() => null, {displayName: `${this.getName()}.${componentName}`})
+
+        return Promise.resolve([
+            {
+                path: requestURL.pathname,
+                component,
+                exact: true
+            }
+        ])
+    }
+
+    /**
+     * TODO: update comment to make it clearer that beforeRouteMatch is also called on the client side
+     * This method is used on the server during the rendering pipeline. It's provided a list of all the routes that your application
+     * is configured with, including those defined in the base application and those added by all the extensions. You can use this
+     * method to modify these routes in any way you want, but you must return an array of routes as a result.
+     */
+    beforeRouteMatch({ allRoutes, locals }: BeforeRouteMatchParams): RouteProps[] {
+        const routes = [...allRoutes]
+        const index = routes.findIndex(
+            (route) => route.component?.displayName?.includes(this.getName())
+        )
+    
+        if (index === -1) return routes
+    
+        // Extract and remove the matched route
+        const [route] = routes.splice(index, 1)
+        const componentName = route.component.displayName.split(".").pop()
+    
+        if (!componentName) return routes
+    
+        const component = routes.find((_route) =>
+            _route.component?.displayName?.includes(componentName)
+        )?.component
+    
+        if (!component) {
+            throw new Error(`Could not find component with displayName "${componentName}"`)
+        }
+    
+        route.component = routeComponent(component, true, locals)
+    
+        return [route, ...routes]
+    }
+
+    getComponentMap(): ComponentMap {
+        // During deserialize each serialized route in this extension MUST return a component.
+        // This implementation returns an object where the keys are component names in resourceTypeToComponentMap
+        // and the values are placeholder components with the displayName
+        const {resourceTypeToComponentMap} = this.getConfig()
+        return Object.values(resourceTypeToComponentMap).reduce((acc: ComponentMap, name) => {
+            acc[`${this.getName()}.${name}`] = Object.assign(() => null, {displayName: `${this.getName()}.${name}`})
```

Comment:
Yes, the component map in this case (SEO extension) returns _placeholder_ components. This may be obvious, but I thought it's quite eye opening that we're supposed to return placeholder components here :)